test/solver3d/generate_source_terms.py

```diff
@@ -0,0 +1,263 @@
+    return res_elev, res_u, res_v, cori_u, cori_v
+
+
+u_3d, v_3d, u_2d, v_2d = split_velocity(u, v, elev, bath)
+w = evaluate_w(elev, u, v, bath)
+rho, baroc_head = evaluate_baroclinicity(elev, temp, salt)
+
+mom_source_x, mom_source_y, int_pg_x, int_pg_y, adv_u, adv_v, adv_w_u, adv_w_v, cori_u, cori_v = evaluate_mom_source(elev, baroc_head, u, v, w, u_3d, v_3d, bath, f, nu)
+
+vol_source_2d, mom_source_2d_x, mom_source_2d_y, cori_u_2d, cori_v_2d = evaluate_swe_source(elev, u_2d, v_2d, bath, f, nu, nonlin=True)
+temp_source_3d, adv_t_uv, adv_t_w = evaluate_tracer_source(elev, temp, u, v, w, bath, f, nu)
+
+
+def expr2str(e):
+    if isinstance(e, (numbers.Number, sympy.numbers.Zero)):
+        return 'Constant({:})'.format(e)
+    return str(e)
+
+
+def print_expr(name, *expr, dict_fmt=True):
+    if len(expr) == 1:
+        expr_str = expr2str(expr[0])
+    else:
+        if all(isinstance(e, numbers.Number) for e in expr):
+            comp_str = ', '.join([str(e) for e in expr])
+            expr_str = 'Constant((' + comp_str + '))'
+        else:
+            comp_str = ', '.join([expr2str(e) for e in expr])
+            expr_str = 'as_vector((' + comp_str + '))'
+    if dict_fmt:
+        print("    out['{:}'] = {:}".format(name, expr_str))
+    else:
+        print("    {:} = {:}".format(name, expr_str))
+
+
+def to_2d_coords(expr):
+    if isinstance(expr, numbers.Number):
+        return expr
+    return expr.subs(x, x_2d).subs(y, y_2d)
+
+
+def print_ufl_expressions():
+    """
+    Print python expressions
+    """
+    print_expr('elev_2d', to_2d_coords(elev))
+    print_expr('uv_full_3d', u, v, 0)
+    print_expr('uv_2d', to_2d_coords(u_2d), to_2d_coords(v_2d))
+    print_expr('uv_dav_3d', u_2d, v_2d, 0)
+    print_expr('uv_3d', u_3d, v_3d, 0)
+    print_expr('w_3d', 0, 0, w)
+    print_expr('temp_3d', temp)
+    print_expr('density_3d', rho)
+    print_expr('baroc_head_3d', baroc_head)
+    print_expr('int_pg_3d', int_pg_x, int_pg_y, 0)
+
+    print_expr('vol_source_2d', to_2d_coords(vol_source_2d))
+    print_expr('mom_source_2d', to_2d_coords(mom_source_2d_x), to_2d_coords(mom_source_2d_y))
+    print_expr('mom_source_3d', mom_source_x, mom_source_y, 0)
+    print_expr('temp_source_3d', temp_source_3d)
+
+
+def print_latex_expressions():
+    """
+    Print expressions in Latex for publications
+    """
+    _x, _y, _z = sympy.symbols('x y z')
+    _lx, _ly = sympy.symbols('L_x L_y')
+    _depth = sympy.symbols('h')
+    _eos_alpha, _eos_t0, _eos_s0 = sympy.symbols('alpha T_0 S_0')
+    _eos_beta = 0
+    _rho0 = sympy.symbols('rho_0', positive=True)
+    _salt0 = sympy.symbols('S_a', positive=True)
+    _g = sympy.symbols('g', positive=True)
+
+    def for_latex(e):
+        o = e.subs(x, _x).subs(y, _y).subs(z, _z).subs(lx, _lx).subs(ly, _ly).subs(depth, _depth).subs(g, _g)
+        o = o.subs(eos_alpha, _eos_alpha).subs(eos_beta, _eos_beta).subs(eos_t0, _eos_t0).subs(eos_s0, _eos_s0).subs(rho0, _rho0).subs(salt0, _salt0)
```

Comment:
This could all be combined in one dictionary: `e.subs({x: _x, y: _y, ... salt0: _salt0})`

Reply:
Yes, that's nicer. Updated now.